File: backend/core/utils/simple_crypto.py
"""
Simple Crypto Utilities for Python
==================================
Simple and reliable encryption/decryption for environment variables
"""

import os
import base64
import logging

logger = logging.getLogger(__name__)


def decrypt_value(encrypted_text: str) -> str:
    """
    Simple decrypt function
    """
    if not encrypted_text or not encrypted_text.startswith('ENC_'):
        return encrypted_text
    
    try:
        encoded = encrypted_text.replace('ENC_', '')
        unreversed = encoded[::-1]  # Reverse string
        decoded = base64.b64decode(unreversed).decode('utf-8')
        
        return decoded
    except Exception as e:
        logger.error(f"Decryption failed: {e}")
        return encrypted_text  # Return original if decryption fails


def get_decrypted_env(key: str, default_value: str = '') -> str:
    """
    Get decrypted environment variable
    """
    value = os.getenv(key, default_value)
    return decrypt_value(value)


def get_decrypted_oauth_config() -> dict:
    """
    Get all decrypted OAuth config
    """
    return {
        'NEXTAUTH_SECRET': get_decrypted_env('NEXTAUTH_SECRET'),
        'GOOGLE_CLIENT_ID': get_decrypted_env('GOOGLE_CLIENT_ID'),
        'GOOGLE_CLIENT_SECRET': get_decrypted_env('GOOGLE_CLIENT_SECRET'),
    }


def get_decrypted_api_keys() -> dict:
    """
    Get decrypted API keys
    """
    return {
        'TAVILY_API_KEY': get_decrypted_env('TAVILY_API_KEY'),
        'GOOGLE_MAPS_API_KEY': get_decrypted_env('GOOGLE_MAPS_API_KEY'),
    }


def safe_log_value(key: str, value: str) -> str:
    """
    Utility to safely log encrypted values (shows only prefix)
    """
    if value and value.startswith('ENC_'):
        return f"{key}: [ENCRYPTED - {value[:10]}...]"
    return f"{key}: [PLAIN - {'[SET]' if value else '[EMPTY]'}]"


# Test function
def test_decryption():
    """
    Test decryption with sample values
    """
    test_values = {
        'GOOGLE_CLIENT_ID': 'ENC_=Umclh2Xkl2X05WZpx2YfVGbn92bn9lc19We',
        'GOOGLE_CLIENT_SECRET': 'ENC_lJXZo9FdlJ3YlN3X05WZpx2YfVGbn92bn9lc19We',
        'NEXTAUTH_SECRET': 'ENC_0NWZq9mcw1ybtVGZtUmbvx2YtEWay9Gd1FWL0IDMy0SeltWL0VmcjV2ctgGd1FGd4VmbtwWYu9Wa0F2Y1RWZ'
    }
    
    print("🧪 Testing decryption:")
    for key, encrypted_value in test_values.items():
        decrypted = decrypt_value(encrypted_value)
        print(f"✅ {key}: {decrypted}")


if __name__ == "__main__":
    test_decryption()

File: frontend/src/lib/simple-crypto.ts
/**
 * Безопасно логирует значение (скрывает чувствительные данные)
 * @param key - Ключ для логирования
 * @param value - Значение для логирования
 * @returns Строка для безопасного логирования
 */
export function safeLogValue(key: string, value: string): string {
  if (!value) {
    return `${key}: [EMPTY]`;
  }
  
  if (value.startsWith('ENC_')) {
    return `${key}: [ENCRYPTED - ${value.substring(0, 15)}...]`;
  }
  
  return `${key}: [DECRYPTED]`;
}

/**
 * Дешифрует зашифрованное значение
 * @param encryptedText - Зашифрованный текст
 * @returns Дешифрованный текст
 */
export function decryptValue(encryptedText: string): string {
  // 1. Check if value is encrypted (starts with ENC_)
  if (!encryptedText || !encryptedText.startsWith('ENC_')) {
    return encryptedText; // Return as-is if not encrypted
  }

  try {
    // 2. Remove ENC_ prefix
    const encoded = encryptedText.replace('ENC_', '');

    // 3. Reverse the string back
    const unreversed = encoded.split('').reverse().join('');

    // 4. Decode from Base64
    const decoded = Buffer.from(unreversed, 'base64').toString('utf8');

    return decoded;
  } catch (error) {
    console.error('Decryption failed:', error);
    return encryptedText; // Return original if decryption fails
  }
}

/**
 * Получает дешифрованную переменную окружения
 * @param key - Ключ переменной
 * @param defaultValue - Значение по умолчанию
 * @returns Дешифрованное значение
 */
export function getDecryptedEnv(key: string, defaultValue: string = ''): string {
  const value = process.env[key] || defaultValue;
  return decryptValue(value);
}

/**
 * Получает расшифрованную конфигурацию OAuth
 * @returns Расшифрованная конфигурация OAuth
 */
export function getDecryptedOAuthConfig(): any {
  // Логируем переменные для диагностики
  console.log('[Constants] Raw environment variables:');
  console.log(`  ${safeLogValue('NEXTAUTH_SECRET', process.env.NEXTAUTH_SECRET || '')}`);
  console.log(`  ${safeLogValue('GOOGLE_CLIENT_ID', process.env.GOOGLE_CLIENT_ID || '')}`);
  console.log(`  ${safeLogValue('GOOGLE_CLIENT_SECRET', process.env.GOOGLE_CLIENT_SECRET || '')}`);

  // Получаем дешифрованную конфигурацию
  const decryptedConfig = {
    NEXTAUTH_SECRET: getDecryptedEnv('NEXTAUTH_SECRET'),
    GOOGLE_CLIENT_ID: getDecryptedEnv('GOOGLE_CLIENT_ID'),
    GOOGLE_CLIENT_SECRET: getDecryptedEnv('GOOGLE_CLIENT_SECRET'),
  };

  // Логируем финальную конфигурацию
  console.log('[Constants] Final AUTH_CONFIG:');
  console.log(`  NEXTAUTH_SECRET: ${decryptedConfig.NEXTAUTH_SECRET ? '[DECRYPTED]' : '[EMPTY]'}`);
  console.log(`  GOOGLE_CLIENT_ID: ${decryptedConfig.GOOGLE_CLIENT_ID ? '[DECRYPTED]' : '[EMPTY]'}`);
  console.log(`  GOOGLE_CLIENT_SECRET: ${decryptedConfig.GOOGLE_CLIENT_SECRET ? '[DECRYPTED]' : '[EMPTY]'}`);

  // Логируем первые символы для проверки
  if (decryptedConfig.GOOGLE_CLIENT_ID) {
    console.log(`  GOOGLE_CLIENT_ID preview: ${decryptedConfig.GOOGLE_CLIENT_ID.substring(0, 20)}...`);
  }
  if (decryptedConfig.GOOGLE_CLIENT_SECRET) {
    console.log(`  GOOGLE_CLIENT_SECRET length: ${decryptedConfig.GOOGLE_CLIENT_SECRET.length}`);
  }

  return decryptedConfig;
}